### PR TITLE
Filter out compiler generated types for `Add-Type -PassThru`

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Only return types that are not auto-generated by compiler with `Add-Type -PassThru`.

An improvement was made to `Microsoft.CodeAnalysis.CSharp` to allow compiler to generate private helper class sometimes for performance reason (see an example below). Passing through all types from the assembly cause one of our tests to fail:

https://dev.azure.com/powershell/PowerShell/_build/results?buildId=108821&view=logs&j=77295986-da6b-5a15-d60e-dfb16acd57ef&t=a58f8c58-10a4-5f19-f6fe-df1d6acdd5db&l=685

This PR makes `-PassThru` to only return types that are not compiler generated. It's technically a breaking change, but should be in "Bucket 3 - gray area".

```
$classDefinition = @'
using System;
using System.Management.Automation;
using System.Management.Automation.Runspaces;
using System.Threading;
namespace StackTest {
    public class StackDepthTest {
        public static PowerShell ps;
        public static int size = 512 * 1024;
        public static void CauseError() {
            Thread t = new Thread(RunPS, size);
            t.Start();
            t.Join();
        }

        public static void RunPS() {
            InitialSessionState iss = InitialSessionState.CreateDefault2();
            iss.ThreadOptions = PSThreadOptions.UseCurrentThread;
            ps = PowerShell.Create(iss);
            ps.AddScript("function recurse { recurse }; recurse").Invoke();
        }

        public static void GetPSError() {
            if ( ps.Streams.Error.Count > 0) {
                throw ps.Streams.Error[0].Exception.InnerException;
            }
        }
    }
}
'@

Add-Type -PassThru -TypeDefinition $classDefinition

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    StackDepthTest                           System.Object
False    False    <>O                                      System.Object
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively

### CODE REVIEW

**Review by adityapatwardhan — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@SeeminglyScience Can you please take a look and see if this will cause any problem to VSCode Extension?

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience Can you please take a look and see if this will cause any problem to VSCode Extension?

I think all is good on the VSCode side 🎉

Though, I can imagine someone generating p/invokes as internal and doing `(Add-Type -Passthru)[0]::PInvokeName()` or something like that to avoid polluting type resolution.

Would definitely be slower, but maybe worth considering looking for `CompilerGeneratedAttribute` decorations as a filter

**Comment by iSazonov** [comment]
> A breaking change is made to new versions of `Microsoft.CodeAnalysis.CSharp` to make the compiler generate private helper class sometimes (see an example below).

Looks like a bug. Why should we hide it?

The fix looks like very sensitive breaking change in the cmdlet.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by SeeminglyScience** [comment]
> > A breaking change is made to new versions of `Microsoft.CodeAnalysis.CSharp` to make the compiler generate private helper class sometimes (see an example below).
> 
> Looks like a bug. Why should we hide it?
> 
> The fix looks like very sensitive breaking change in the cmdlet.

Roslyn has always created private helper types in certain conditions, typically for performance reasons. I don't think they would consider this change breaking, they expect all tooling that reads types to ignore anything decorated with `CompilerGeneratedAttribute` when the exclusion of them is required.

**Comment by iSazonov** [comment]
> > > A breaking change is made to new versions of `Microsoft.CodeAnalysis.CSharp` to make the compiler generate private helper class sometimes (see an example below).
> > 
> > 
> > Looks like a bug. Why should we hide it?
> > The fix looks like very sensitive breaking change in the cmdlet.
> 
> Roslyn has always created private helper types in certain conditions, typically for performance reasons. I don't think they would consider this change breaking, they expect all tooling that reads types to ignore anything decorated with `CompilerGeneratedAttribute` when the exclusion of them is required.

Maybe they lost the attribute for member we see? :-)

And your comment raises the same question - why should we hide if they don't?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs [hunk 1]
```diff
@@ -891,7 +892,14 @@ private IEnumerable<PortableExecutableReference> GetPortableExecutableReferences
 
         private void WriteTypes(Assembly assembly)
         {
-            WriteObject(assembly.GetTypes(), true);
+            foreach (Type type in assembly.GetTypes())
+            {
+                // We only write out types that are not auto-generated by compiler.
+                if (type.GetCustomAttribute<CompilerGeneratedAttribute>() is null)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:898]: Can we use:
```suggestion
                if (type.IsDefined(typeof(CompilerGeneratedAttribute), inherit: false))
```

**Review by SeeminglyScience — APPROVED** [review]
LGTM! ❤️

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:898]: I don't think there is much difference here. I will keep it as is.

**Comment by daxian-dbw** [comment]
> Maybe they lost the attribute for member we see? :-)
And your comment raises the same question - why should we hide if they don't?

Nope, nothing is hidden. It's an improvement to the emitted code.
- The assembly generated from the same source code in PS 7.2.6 doesn't do any optimization around wrapping `RunPS` to a delegate -- every time `CauseError` is called, the static method `RunPS` will be wrapped to a new delegate.
- With the improvement in the new `Microsoft.CodeAnalysis.CSharp`, a helper class is generated to hold the delegate, so the delegate gets created only once.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs (hunk 1 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:898]: GetCustomAttribute() does extra allocations at least.

**Review by SteveL-MSFT — APPROVED** [review]

**Pull request merged**